File: 213-house-robber-ii/house-robber-ii.cpp
class Solution {
public:
int find(int i,vector<int>&v,vector<int>&nums){
      if(i<0) return 0;
    if(i==0) return nums[i];
  
    if(v[i]!=-1) return v[i];
    int take =find(i-2,v,nums) + nums[i];
    int nottake=find(i-1,v,nums);
    
     return v[i]=max(take,nottake);
}
    int rob(vector<int>& nums) {
        if(nums.size()==1) return nums[0];
        vector<int>v1,v2;
        vector<int>v(nums.size(),-1);
        vector<int>s(nums.size(),-1);
        for(int i=0;i<nums.size();i++){
            if(i!=0)v1.push_back(nums[i]);
            if(i!=nums.size()-1) v2.push_back(nums[i]);
        }
        int i=v1.size()-1;
        return max(find(i,v,v1),find(i,s,v2));
    }
};

File: 213-house-robber-ii/house-robber-ii.java
class Solution {
     int find(int i,int[]dp,int[] h){
        if(i==0) return h[0];
        if(i==1) return Math.max(h[0],h[1]);
        if(dp[i]!=-1)return dp[i];
        int take=find(i-2,dp,h)+h[i];
        int nottake=find(i-1,dp,h);
        return dp[i]=Math.max(take,nottake);
    }
    public int rob(int[] nums) {
        
        int n=nums.length;
        if(n==1) return nums[0];
        if(n==2) return Math.max(nums[0],nums[1]);
        int a[]=new int[n-1];
        int b[]=new int[n-1];
        int dp1[]=new int[n+1];
        int dp2[]=new int[n+1];
        Arrays.fill(dp1,-1);
        Arrays.fill(dp2,-1);
        int j=0,k=0;
        for(int i=0;i<n;i++){
            if(i!=n-1){
                a[j++]=nums[i];
            }
            if(i!=0){
                b[k++]=nums[i];
            }
        }
        return Math.max(find(n-2,dp1,a),find(n-2,dp2,b));
    }
}